src/core/public/workspace/workspaces_service.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BehaviorSubject, combineLatest } from 'rxjs';
+import { isEqual } from 'lodash';
+
+import { CoreService, WorkspaceAttribute } from '../../types';
+
+type WorkspaceObject = WorkspaceAttribute & { readonly?: boolean };
+
+interface WorkspaceObservables {
+  /**
+   * Indicates the current activated workspace id, the value should be changed every time
+   * when switching to a different workspace
+   */
+  currentWorkspaceId$: BehaviorSubject<string>;
+
+  /**
+   * The workspace that is derived from `currentWorkspaceId` and `workspaceList`, if
+   * `currentWorkspaceId` cannot be found from `workspaceList`, it will return an error
+   *
+   * This value MUST NOT set manually from outside of WorkspacesService
+   */
+  currentWorkspace$: BehaviorSubject<WorkspaceObject | null>;
+
+  /**
+   * The list of available workspaces. This workspace list should be set by whoever
+   * implemented the workspace functionalities, core workspace module should not be
+   * aware of how to populate the workspace list.
+   */
+  workspaceList$: BehaviorSubject<WorkspaceObject[]>;
+
+  /**
+   * This is a flag which indicates the WorkspacesService module is initialized and ready
+   * for consuming by others. For example, the `workspaceList` has been set, etc
+   */
+  initialized$: BehaviorSubject<boolean>;
```

Comment:
this can only go from `not initialized` to `initialized`, does it needs to be an observable?

Reply:
The intention of making it reactive is, the value changing from `not initialized` to `initialized ` maybe happen at any time, this is unlike reading regular configurations which we get value when application is mount. Using an observable so that the application can react to the value change.